# EDA: Diagnosing Diabetes

In this project, you'll imagine you are a data scientist interested in exploring data that looks at how certain diagnostic factors affect the diabetes outcome of women patients.

You will use your EDA skills to help inspect, clean, and validate the data.

**Note**: This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database) is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration per 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure
- `SkinThickness`: Triceps skinfold thickness
- `Insulin`: 2-Hour serum insulin
- `BMI`: Body mass index
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

Let's get started!

## Initial Inspection

1. First, familiarize yourself with the dataset [here](https://www.kaggle.com/uciml/pima-indians-diabetes-database).

   Look at each of the nine columns in the documentation.
   
   What do you expect each data type to be?

Expected data type for each column:

- `Pregnancies`: 
- `Glucose`: 
- `BloodPressure`: 
- `SkinThickness`: 
- `Insulin`: 
- `BMI`: 
- `DiabetesPedigreeFunction`: 
- `Age`: 
- `Outcome`: 

2. Next, let's load in the diabetes data to start exploring.

   Load the data in a variable called `diabetes_data` and print the first few rows.
   
   **Note**: The data is stored in a file called `diabetes.csv`.

In [1]:
import pandas as pd
import numpy as np

# load in data
diabetes_data = pd.read_csv('diabetes.csv')
print(diabetes_data.head())

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1  


3. How many columns (features) does the data contain?

In [2]:
# print number of columns
print(9)

9


4. How many rows (observations) does the data contain?

In [3]:
# print number of rows
print(len(diabetes_data))

768


## Further Inspection

5. Let's inspect `diabetes_data` further.

   Do any of the columns in the data contain null (missing) values?

In [4]:
# find whether columns contain null values
print(diabetes_data.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


6. If you answered no to the question above, not so fast!

   While it's technically true that none of the columns contain null values, that doesn't necessarily mean that the data isn't missing any values.
   
   When exploring data, you should always question your assumptions and try to dig deeper.
   
   To investigate further, calculate summary statistics on `diabetes_data` using the `.describe()` method.

In [5]:
# perform summary statistics
print(diabetes_data.describe())
print(diabetes_data.info())

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000        

7. Looking at the summary statistics, do you notice anything odd about the following columns?

   - `Glucose`
   - `BloodPressure`
   - `SkinThickness`
   - `Insulin`
   - `BMI`

**Your response to question 7**: The minimum value for each of these columns is listed as zero, which doesn't seem correct for living individuals.

8. Do you spot any other outliers in the data?

**Your response to question 8**: 17 pregnancies seems like an outlier, as well as 99 mm for skin thickness, 846 for insulin, and 67.1 for BMI. 

There may be other outliers. And notice how the standard deviation for insulin is so much more than the mean? I think there must be many missing entries for insulin. Look at the first quartile value! It's still zero!

9. Let's see if we can get a more accurate view of the missing values in the data.

   Use the following code to replace the instances of `0` with `NaN` in the five columns mentioned:
   
   ```py
   diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)
   ```

In [6]:
# replace instances of 0 with NaN
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)

10. Next, check for missing (null) values in all of the columns just like you did in Step 5.

    Now how many missing values are there?

In [7]:
# find whether columns contain null values after replacements are made
print(diabetes_data.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


11. Let's take a closer look at these rows to get a better idea of _why_ some data might be missing.

    Print out all the rows that contain missing (null) values.

In [8]:
# print rows with missing values
print(len(diabetes_data[diabetes_data.isnull().any(axis=1)]))

376


In [9]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the first 50 rows. (1 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][:50]))

    Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             6    148.0           72.0           35.0      NaN  33.6   
1             1     85.0           66.0           29.0      NaN  26.6   
2             8    183.0           64.0            NaN      NaN  23.3   
5             5    116.0           74.0            NaN      NaN  25.6   
7            10    115.0            NaN            NaN      NaN  35.3   
9             8    125.0           96.0            NaN      NaN   NaN   
10            4    110.0           92.0            NaN      NaN  37.6   
11           10    168.0           74.0            NaN      NaN  38.0   
12           10    139.0           80.0            NaN      NaN  27.1   
15            7    100.0            NaN            NaN      NaN  30.0   
17            7    107.0           74.0            NaN      NaN  29.6   
21            8     99.0           84.0            NaN      NaN  35.4   
22            7    196.0           90.0            

In [10]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the next 50 rows. (2 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][50:100]))

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
86            13    106.0           72.0           54.0      NaN  36.6   
89             1    107.0           68.0           19.0      NaN  26.5   
90             1     80.0           55.0            NaN      NaN  19.1   
93             4    134.0           72.0            NaN      NaN  23.8   
96             2     92.0           62.0           28.0      NaN  31.6   
100            1    163.0           72.0            NaN      NaN  39.0   
101            1    151.0           60.0            NaN      NaN  26.1   
102            0    125.0           96.0            NaN      NaN  22.5   
104            2     85.0           65.0            NaN      NaN  39.6   
106            1     96.0          122.0            NaN      NaN  22.4   
113            4     76.0           62.0            NaN      NaN  34.0   
115            4    146.0           92.0            NaN      NaN  31.2   
116            5    124.0           74

In [11]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the next 50 rows. (3 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][100:150]))

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
193           11    135.0            NaN            NaN      NaN  52.3   
194            8     85.0           55.0           20.0      NaN  24.4   
196            1    105.0           58.0            NaN      NaN  24.3   
200            0    113.0           80.0           16.0      NaN  31.0   
201            1    138.0           82.0            NaN      NaN  40.1   
202            0    108.0           68.0           20.0      NaN  27.3   
205            5    111.0           72.0           28.0      NaN  23.9   
207            5    162.0          104.0            NaN      NaN  37.7   
209            7    184.0           84.0           33.0      NaN  35.5   
210            2     81.0           60.0           22.0      NaN  27.7   
211            0    147.0           85.0           54.0      NaN  42.8   
212            7    179.0           95.0           31.0      NaN  34.2   
218            5     85.0           74

In [12]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the next 50 rows. (4 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][150:200]))

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
284            2    108.0           80.0            NaN      NaN  27.0   
294            0    161.0           50.0            NaN      NaN  21.9   
299            8    112.0           72.0            NaN      NaN  23.6   
300            0    167.0            NaN            NaN      NaN  32.3   
303            5    115.0           98.0            NaN      NaN  52.9   
304            3    150.0           76.0            NaN      NaN  21.0   
310            6     80.0           66.0           30.0      NaN  26.2   
314            7    109.0           80.0           31.0      NaN  35.9   
317            3    182.0           74.0            NaN      NaN  30.5   
319            6    194.0           78.0            NaN      NaN  23.5   
321            3    112.0           74.0           30.0      NaN  31.6   
322            0    124.0           70.0           20.0      NaN  27.4   
324            2    112.0           75

In [13]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the next 50 rows. (5 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][200:250]))

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
406            4    115.0           72.0            NaN      NaN  28.9   
407            0    101.0           62.0            NaN      NaN  21.9   
408            8    197.0           74.0            NaN      NaN  25.9   
410            6    102.0           90.0           39.0      NaN  35.7   
416            1     97.0           68.0           21.0      NaN  27.2   
417            4    144.0           82.0           32.0      NaN  38.5   
418            1     83.0           68.0            NaN      NaN  18.2   
423            2    115.0           64.0           22.0      NaN  30.8   
426            0     94.0            NaN            NaN      NaN   NaN   
430            2     99.0            NaN            NaN      NaN  22.2   
433            2    139.0           75.0            NaN      NaN  25.6   
434            1     90.0           68.0            8.0      NaN  24.5   
435            0    141.0            N

In [14]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the next 50 rows. (6 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][250:300]))

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
512            9     91.0           68.0            NaN      NaN  24.2   
513            2     91.0           62.0            NaN      NaN  27.3   
517            7    125.0           86.0            NaN      NaN  37.6   
518           13     76.0           60.0            NaN      NaN  32.8   
522            6    114.0            NaN            NaN      NaN   NaN   
523            9    130.0           70.0            NaN      NaN  34.2   
524            3    125.0           58.0            NaN      NaN  31.6   
525            3     87.0           60.0           18.0      NaN  21.8   
529            0    111.0           65.0            NaN      NaN  24.6   
531            0    107.0           76.0            NaN      NaN  45.3   
533            6     91.0            NaN            NaN      NaN  29.8   
535            4    132.0            NaN            NaN      NaN  32.9   
536            0    105.0           90

In [15]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the next 50 rows. (7 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][300:350]))

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
619            0    119.0            NaN            NaN      NaN  32.4   
621            2     92.0           76.0           20.0      NaN  24.2   
622            6    183.0           94.0            NaN      NaN  40.8   
624            2    108.0           64.0            NaN      NaN  30.8   
626            0    125.0           68.0            NaN      NaN  24.7   
627            0    132.0           78.0            NaN      NaN  32.4   
628            5    128.0           80.0            NaN      NaN  34.6   
629            4     94.0           65.0           22.0      NaN  24.7   
630            7    114.0           64.0            NaN      NaN  27.4   
632            2    111.0           60.0            NaN      NaN  26.2   
634           10     92.0           62.0            NaN      NaN  25.9   
635           13    104.0           72.0            NaN      NaN  31.2   
636            5    104.0           74

In [16]:
# I find that my output looks best if I print only about 50 rows of data at a time. Here are the final rows. (8 of 8)

# print rows with missing values
print((diabetes_data[diabetes_data.isnull().any(axis=1)][350:]))

     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
720            4     83.0           86.0           19.0      NaN  29.3   
724            1    111.0           94.0            NaN      NaN  32.8   
725            4    112.0           78.0           40.0      NaN  39.4   
727            0    141.0           84.0           26.0      NaN  32.4   
728            2    175.0           88.0            NaN      NaN  22.9   
729            2     92.0           52.0            NaN      NaN  30.1   
731            8    120.0           86.0            NaN      NaN  28.4   
734            2    105.0           75.0            NaN      NaN  23.3   
735            4     95.0           60.0           32.0      NaN  35.4   
737            8     65.0           72.0           23.0      NaN  32.0   
739            1    102.0           74.0            NaN      NaN  39.5   
743            9    140.0           94.0            NaN      NaN  32.7   
746            1    147.0           94

12. Go through the rows with missing data. Do you notice any patterns or overlaps between the missing data?

**Your response to question 12**: Yes, there is frequent overlap between missing SkinThickness and Insulin values.

13. Next, take a closer look at the data types of each column in `diabetes_data`.

    Does the result match what you would expect?

In [17]:
# print data types using .info() method
print(diabetes_data.dtypes)
print(diabetes_data.info())

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                      object
dtype: object
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Out

14. To figure out why the `Outcome` column is of type `object` (string) instead of type `int64`, print out the unique values in the `Outcome` column.

In [18]:
# print unique values of Outcome column
print(diabetes_data['Outcome'].unique())

['1' '0' 'O']


15. How might you resolve this issue?

**Your response to question 15**: You've got some Outcomes recorded with an upper-case O rather than 0 (zero).

In [19]:
diabetes_data = diabetes_data.replace({'Outcome': {'1': 1, '0': 0, 'O': 0}})

In [20]:
# print unique values of Outcome column
print(diabetes_data.dtypes)

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


## Next Steps:

16. Congratulations! In this project, you saw how EDA can help with the initial data inspection and cleaning process. This is an important step as it helps to keep your datasets clean and reliable.

    Here are some ways you might extend this project if you'd like:
    - Use `.value_counts()` to more fully explore the values in each column.
    - Investigate other outliers in the data that may be easily overlooked.
    - Instead of changing the `0` values in the five columns to `NaN`, try replacing the values with the median or mean of each column.